lib/bolero-engine/src/rng.rs

```diff
@@ -97,9 +115,10 @@ where
 
         let mut state = RngState::new(self.seed, self.max_len, options);
 
+        let start_time = Instant::now();
         let mut valid = 0;
         let mut invalid = 0;
-        while valid < self.iterations {
+        while valid < self.iterations && start_time.elapsed() < self.test_time {
```

Comment:
Maybe as an optimization we can avoid querying the clock if `test_time == Duration::MAX`.

I don't know if you've had time to profile this change, but I'd imagine the `elapsed` call isn't free. This ends up mattering for harnesses that are super cheap to run. I wonder if it might be better to spawn a background thread that just sleeps for this amount of time and sets an `AtomicBool` when it wakes up? Or maybe it's as simple as only querying every N iterations (where N is like 2-5)?

Reply:
TL;DR: I would be very surprised if any real-world fuzzer were to be slowed down in any noticeable manner by this change, and I think the early optimization would bring more complexity than it's worth.

A few performance notes:
- Back in 2010, [Solaris executed clock_gettime in 0.6μs](https://stackoverflow.com/questions/3218855/which-one-faster-getimeofday-or-clock-gettime). This was probably before all the optimizations I'm going to discuss below.
- On Linux, rustc uses clock_gettime. It [is implemented here](https://github.com/torvalds/linux/blob/d206a76d7d2726f3b096037f2079ce0bd3ba329b/lib/vdso/gettimeofday.c#L227-L269) (this is the version that does not take advantage of any architecture-specific optimization). This implementation, for most clocks, is based on a shared memory page, that Linux updates at some regularity. Then, if high-performance counters are required, it uses `rdtsc` to make things more precise. This means that computing the current time does not require a syscall and can be very fast
- Internet browsers, at some point (at least that's what I was told at Mozilla in 2016), were hogged by `gettimeofday`, where a significant amount of time was used there, for reasons that were not possible to avoid due to the specification. This pushed for lots of performance improvements in time handling, and in particular this VDSO-only computation of time that does not require a syscall
- With the default 1000 iterations, even with the performance of solaris in 2010, the total overhead would be 0.6ms, which is not human-noticeable.
- For random fuzzing of functions that take less than 10μs (100k execs/s) to run the overhead would have been noticeable (6%) back in 2010, but do these actually exist? I don't think I actually have fuzzers that run more than 1k execs/s, except for a few trivial ones that had gotten to basically full feature coverage in a few minutes and so are not worth optimizing for. And anyway I'd expect random fuzzing to be a very marginal use case anyway, as it's so much worse than coverage-based fuzzing :)

Reply:
All good points 🙂